lib/puppet/state_machine/machine_builder.rb

```diff
@@ -0,0 +1,240 @@
+require 'puppet/state_machine/state'
+
+module Puppet
+  class StateMachine
+    # The MachineBuilder class provides a DSL for implementing the states and transitions of a state machine,
+    # and validating that the machine is well behaved.
+    #
+    # ## Composing state machines
+    #
+    # Multiple state machines can be composed into a single unified state machine with the {#compose} method.
+    # When state machines are composed the state names and transitions are namespaced based on the originating
+    # state machine.
+    #
+    # ### Final states
```

Comment:
Should most of this doc be in `state_machine.rb` since `StateMachine::build` is the top-level way to build a machine with this class as an implementation detail?